Phase05

rahaahmadi opened the pull request.

Melikad committed:
Add files via upload

create solution and projects
reference added

Melikad committed:
Add files via upload

solution and projects created
reference added

rahaahmadi committed:
Add files via upload

rahaahmadi committed:
Delete phase05/Solution1 directory

Melikad committed:
DocsFIleReaderTest implemented

rahaahmadi committed:
Add files via upload

rahaahmadi committed:
Delete File1

rahaahmadi committed:
Delete File2

rahaahmadi committed:
Delete File3

rahaahmadi committed:
Add files

rahaahmadi committed:
Add files

rahaahmadi committed:
Add files

rahaahmadi committed:
Delete Phase5/TestFiles directory

Melikad committed:
FilterApplier(Test) implemented (not completely)

rahaahmadi committed:
Updated

rahaahmadi committed:
Cleaned FilterApplier.cs

rahaahmadi committed:
Create FullTextSearch class and test (not completely implemented)

rahaahmadi committed:
Create interface for FilterApplier

rahaahmadi committed:
Commit

rahaahmadi committed:
Refactor + write test for get not existing word condition

rahaahmadi committed:
Handle IO exception

rahaahmadi committed:
Updated tokenizer

rahaahmadi committed:
Files added

Melikad committed:
FilterApplier implemented completely
bugs fixed

rahaahmadi committed:
FullTextSearchTest.cs implemented

rahaahmadi committed:
Merge remote-tracking branch 'origin/Phase05' into Phase05

Melikad committed:
filter test changed into theory with 3 inputs
filter splitCommand method name changed
buges fixed

Melikad committed:
Merge remote-tracking branch 'origin/Phase05' into Phase05

rahaahmadi committed:
Main implemented

rahaahmadi committed:
Merge remote-tracking branch 'origin/Phase05' into Phase05

rahaahmadi committed:
Updated

rahaahmadi committed:
InvertedIndex updated

rahaahmadi committed:
FullTextSearchTest.cs updated

rahaahmadi committed:
Changed wordsMap default value

rahaahmadi committed:
Refactor FilterApplier

rahaahmadi committed:
Add files

rahaahmadi committed:
Refactor FilterApplier.cs

rahaahmadi committed:
Add files

rahaahmadi committed:
Create interface for FullTextSearch.cs

rahaahmadi committed:
Fixed FullTextSearchTest.cs bug

rahaahmadi committed:
Commit

rahaahmadi committed:
Delete .idea/.idea.phase05.dir/.idea directory

rahaahmadi committed:
Delete Phase5/.idea/.idea.Phase5/.idea directory

rahaahmadi committed:
Delete Phase5/Search/bin/Debug/net5.0 directory

rahaahmadi committed:
Delete Phase5/Search/obj directory

rahaahmadi committed:
Delete Phase5.sln.DotSettings.user

rahaahmadi committed:
Create .gitignore

rahaahmadi committed:
Delete Phase5/Search/EnglishData directory

rahaahmadi committed:
Delete Phase5/SearchTest/bin/Debug/net5.0 directory

rahaahmadi committed:
Delete Phase5/SearchTest/obj directory

rahaahmadi committed:
Delete .gitignore

rahaahmadi committed:
Make file address constant

rahaahmadi committed:
Merge remote-tracking branch 'origin/Phase05' into Phase05

rahaahmadi committed:
Update Program.cs

rahaahmadi committed:
Update Program.cs

rahaahmadi committed:
Code cleaned

nedataghizadeh79 commented on Phase5/Search/FilterApplier.cs:
```diff
@@ -0,0 +1,76 @@
+            return wordsDocs;
+        }
+    }
+}
```
بنظر این کلاس خیلی طولانیه میتونید تو چندتا کلاس پیاده سازیش کنید

nedataghizadeh79 replied:
مثلا برای نگهداری هش ست ها کلاس جدا بزنید 
برای اشتراک یه کلاس جدا 
برای اجتماع هم ی کلاس جدا دیگه اینجوری تمیز تره

zkhotanlou commented on Phase5/Search/FilterApplier.cs:
```diff
@@ -0,0 +1,76 @@
+
+        }
+        
+        private void SplitCommandWordsBySign(string[] commandWords)
```
This method doesn't relate to FilterApplier class, you can create another class for it

zkhotanlou commented on Phase5/Search/FilterApplier.cs:
```diff
@@ -0,0 +1,76 @@
+            var wordsDocs = new HashSet<string>();
+            foreach (var docs in noSignWords.Select(word => _invertedIndex.GetWordDocs(word)))
+            {
+                if (!wordsDocs.Any())
```
It's better to initialize this Set when you declare it, then this if clause can be deleted

zkhotanlou commented on Phase5/Search/FullTextSearch.cs:
```diff
@@ -0,0 +1,42 @@
+            return _filterApplier.Filter(commandWords);
+        }
+
+        private void LoadDocs()
```
As I understand this part of code is initializing invertedIndex, so it's better to move it in that class or create a new one just for filling inverteIndexMap.

zkhotanlou commented on Phase5/Search/Tokenizer.cs:
```diff
@@ -0,0 +1,23 @@
+    public class Tokenizer : ITokenizer
+    {
+        public HashSet<string> Tokenize(string doc)
+        {
```
Sounds that you can implement this method simpler, It's little hard to understand what happened here !!

zkhotanlou commented on Phase5/Search/Tokenizer.cs:
```diff
@@ -0,0 +1,23 @@
+            {
+                var word = match.Value;
+                if (word.Length > 1)
+                    ans.Add(word.ToLower());
```
it's better to make whole of content lower case at the beginning of method so you will have less overhead of calling this method in every iteration of foreach loop.

nedataghizadeh79 commented on Phase5/Search/Tokenizer.cs:
```diff
@@ -0,0 +1,23 @@
+        public HashSet<string> Tokenize(string doc)
+        {
+            var ans = new HashSet<string>();
+            var regex = new Regex("[a-zA-Z]+");
```
میتوستید از رجکس[\\W]+  
 هم استفاده کنید
بنظرم جذابتره

rahaahmadi committed:
Removed useless method

rahaahmadi committed:
Refactor Tokenizer.cs

rahaahmadi committed:
Creat a class for query processing

rahaahmadi committed:
Refactor Tokenizer.cs

rahaahmadi committed:
Rename variable

hoseinaghaei commented on Phase5/Search/DocsFileReader.cs:
```diff
@@ -0,0 +1,37 @@
+            return filesAddress.ToDictionary(Path.GetFileNameWithoutExtension, File.ReadAllText);
+        }
+
+        private IEnumerable<string> GetAllFiles()
```
شما دارین ادرس همه فایل هارو میگیرین نه خود فایل ها
پس اسم تابع غلطع

hoseinaghaei commented on Phase5/Search/FilterApplier.cs:
```diff
@@ -0,0 +1,51 @@
+            var wordsDocs = new HashSet<string>();
+            foreach (var docs in signWords.Select(word => _invertedIndex.GetWordDocs(word)))
+            {
+                wordsDocs.UnionWith(docs);
```
این قسمت رو میتونید با لینک هم بزنید.
برید راجع 
aggregate
بخونید

hoseinaghaei commented on Phase5/Search/FullTextSearch.cs:
```diff
@@ -0,0 +1,40 @@
+        private readonly IQueryProcessor _queryProcessor;
+        private readonly IFilterApplier _filterApplier;
+
+        public FullTextSearch(IInvertedIndex invertedIndex, IDocsFileReader docsFileReader,ITokenizer tokenizer,IQueryProcessor queryProcessor, IFilterApplier filterApplier)
```
مبتونید هر پارامتر رو توی یک خط بنویسین
اینطوری نمیشه خوندش

hoseinaghaei commented on Phase5/Search/FullTextSearch.cs:
```diff
@@ -0,0 +1,40 @@
+        public HashSet<string> FindCommandResult(string command)
+         {
+            LoadDocs();
+            _queryProcessor.SplitCommandWordsBySign(command);
```
بهتره تابع 
SplitCommandWordsBySign
یک کلاس که شامل اون ۳ تا لیست هست رو برگردونه
بجای اینکه یچیزی تو خودش ذخیره کنه چون ربطی بهش نداره و اسم تابع هم یعنی ی چیزی داره ریترن میشه

hoseinaghaei commented on Phase5/Search/FullTextSearch.cs:
```diff
@@ -0,0 +1,40 @@
+            foreach (var (docId, docContent) in docsMap)
+            {
+                var docWords = _tokenizer.Tokenize(docContent);
+                _invertedIndex.AddDoc(docWords, docId);
```
کلا یادتون باشه اینکه یک کلاس از بیرون ۱۰۰۰ بار صدا زده بشه سربار داره

_invertedIndexبهتره لیست خودتون رو اماده کنید و مستقیما پاسش بدین به

hoseinaghaei commented on Phase5/Search/IQueryProcessor.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface IQueryProcessor
+    { 
+        HashSet<string> PlusCommandWords { get; set; }
```
وقتی تو خود کلاس 
IQueryProcessor
دارین ست میکنین  ۳ تا هش ست رو پس باید 
setter 
اونها 
private
باشه

hoseinaghaei commented on Phase5/Search/Program.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    class Program
+    {
+        static void PrintCommandResult(HashSet<string> commandResult)
```
پنج فاز رفت شما اخر این پرینت و اسکن رو نبردید تو ی اینترفیس دیگه:))

hoseinaghaei commented on Phase5/SearchTest/FilterApplierTest.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            _invertedIndex = Substitute.For<IInvertedIndex>();
+        }
+        private void SetupInvertedIndex()
```
همیشه این ترتیب رو رعایت کنین تا کدتون قشنگ تر شه
اول بالای کلاس فیلد های کانست را ریدانلی رو بذارین
بعد کانستراکتور
بعد توابع پابلیک
و بعدش توابع پرایوت

hoseinaghaei commented on Phase5/SearchTest/FilterApplierTest.cs:
```diff
@@ -0,0 +1,38 @@
+            _invertedIndex.GetWordDocs("xunit").Returns(new HashSet<string>{"File3.txt"});
+        }
+        [Theory]
+        [InlineData(new [] {"microsoft"},new [] {"xunit"},new [] {"hello"}, new string[] {})]
```
تئوری رو طور دیگه ای هم میشه بهش دیتا داد 
بجای 
InlineData
که معمولا برای وقتی که یک ورودی کلا میخوایم بدیم استفاده میشه میتونید از 
MemberData
استفاده کنید.
سرچ کنین ازش استفاده کنین

hoseinaghaei commented on Phase5/SearchTest/FullTextSearchTest.cs:
```diff
@@ -0,0 +1,85 @@
+        public void TestSearchQueryWithOneNoSignWord()
+        {
+            SetupInterfaces();
+            var query = "hello";
```
const string query = "hello";
دلیلش واضحه؟

hoseinaghaei commented on Phase5/SearchTest/FullTextSearchTest.cs:
```diff
@@ -0,0 +1,85 @@
+            var query = "hello +java -sample";
+            _queryProcessor.MinusCommandWords.Returns(new HashSet<string>(){"sample"});
+            _queryProcessor.PlusCommandWords.Returns(new HashSet<string>(){"java"});
+            _queryProcessor.NoSignCommandWords.Returns(new HashSet<string>(){"hello"});
```
setup 
رو مثل اون تابع که بالا گذاشتین ببرین تو ی تابع دیگه وقتی زیاده تعدادش

hoseinaghaei commented on Phase5/SearchTest/FullTextSearchTest.cs:
```diff
@@ -0,0 +1,85 @@
+            Assert.Equal(expectedResult, actualResult);
+        }
+        
+        private void SetupInterfaces()
```
اسم بهتر 
اینترفیس چیه دیگه

hoseinaghaei commented on Phase5/SearchTest/FullTextSearchTest.cs:
```diff
@@ -0,0 +1,85 @@
+        }
+        private void SetupFilterApplier()
+        {
+            _filterApplier.Filter(Arg.Is<HashSet<string>>(x => x.Count == 1),Arg.Is<HashSet<string>>(x => x.Count == 1),Arg.Is<HashSet<string>>(x => x.Count == 1)).Returns(new HashSet<string>(){"File1"});
```
ی چندتا انتر بزنید بفهمیم چی داره میشه:))

hoseinaghaei commented on Phase5/SearchTest/InvertedIndexTest.cs:
```diff
@@ -0,0 +1,38 @@
+        }
+
+        [Fact]
+        public void TestGetExistingWordDocuments()
```
اسم توابع رو به اون شکلی که تو کارگاه گفتم بذارین

hoseinaghaei commented on Phase5/SearchTest/InvertedIndexTest.cs:
```diff
@@ -0,0 +1,38 @@
+        [Fact]
+        public void TestGetExistingWordDocuments()
+        {
+            SetupInvertedIndex();
```
وقتی دارین اینطوری ستاپ میکنین یعنی کارتون مشکل داشته
شما عملا دارین از پیاده سازی تابع 
add
استفاده میکنید

hoseinaghaei commented on Phase5/SearchTest/InvertedIndexTest.cs:
```diff
@@ -0,0 +1,38 @@
+        }
+
+        [Fact]
+        public void TestGetNotExistingWordDocuments()
```
این دوتا تابع رو هم میتونید تئوری کنید

hoseinaghaei reviewed (commented):
افرین در کل خوب بودین اما یکی دوجا رو میشع تغییر داد تا بهتر شه.

hoseinaghaei commented on Phase5/Search/DocsFileReader.cs:
```diff
@@ -0,0 +1,37 @@
+    {
+        private readonly string _filePath;
+
+        public DocsFileReader(string filePath)
```
به نظر شما چرا بهتره ادرس رو تو ورودی تابع 
ReadContent
بگیرین جای کانستراکتور؟

Melikad replied:
:]
توی دو تا تابع استفاده می‌شه بعدش. به جفتش پاس بدیم؟

Parsa2820 reviewed (approved):
<p dir="rtl">
خوب بود. خسته نباشید. کامنتای بقیه رو رفع کنید سعی کنید از namespace هم استفاده کنید.
</p>

rahaahmadi committed:
Rename method

Melikad committed:
comments on FullTextSearch.cs resolved

rahaahmadi committed:
Use aggregate in GetDocs method

Melikad committed:
FullTextSearch.cs comments resolved.
InvertedIndex wordsMap now is fully constructed in FullTextSearch (not using 1000 times calling AddDoc method).

Melikad committed:
Merge remote-tracking branch 'origin/Phase05' into Phase05

Melikad committed:
FullTextSearchTest comment resolved

rahaahmadi committed:
Use theory instead of fact in test + rename method

rahaahmadi committed:
Merge remote-tracking branch 'origin/Phase05' into Phase05

rahaahmadi committed:
Change order of methods (public then private)

rahaahmadi committed:
Created IO interface

rahaahmadi committed:
added method to build inverted index in InvertedIndex.cs + fixed tests

rahaahmadi committed:
used constant variable + deleted useless method

rahaahmadi committed:
Get file address from method parameter instead of constructor

rahaahmadi committed:
Created object in constructor instead of in method

rahaahmadi committed:
updated

rahaahmadi committed:
Removed setter

rahaahmadi committed:
Moved interfaces into another folder
